Connections between patches and regions should be shown

Patryk26g opened the pull request:
**Brief Description of What This PR Does**

This PR adds connections between the patches and regions sides. I increased the margin between the patches and also changed the color of the connections to gray (subject to change) because otherwise it would get a little bit to crowded with all of the connections.
![image](https://github.com/user-attachments/assets/cd167f4c-18b4-487c-a3d6-50a60bacca4d)
![image](https://github.com/user-attachments/assets/151cde1d-3cf8-47cf-9d0c-b6b598cc08a0)
![Screenshot_20250101_131756](https://github.com/user-attachments/assets/b45d58c3-cdd6-4832-8d3e-6377ee0f028c)


**Related Issues**

Closes #3556

**Progress Checklist**

Note: before starting this checklist the PR should be marked as non-draft.

- [x] PR author has checked that this PR works as intended and doesn't
      break existing features:
      https://wiki.revolutionarygamesstudio.com/wiki/Testing_Checklist
      (this is important as to not waste the time of Thrive team
      members reviewing this PR)
- [ ] Initial code review passed (this and further items should not be checked by the PR author)
- [ ] Functionality is confirmed working by another person (see above checklist link)
- [ ] Final code review is passed and code conforms to the 
      [styleguide](https://github.com/Revolutionary-Games/Thrive/blob/master/doc/style_guide.md).

Before merging all CI jobs should finish on this PR without errors, if
there are automatically detected style issues they should be fixed by
the PR author. Merging must follow our
[styleguide](https://github.com/Revolutionary-Games/Thrive/blob/master/doc/style_guide.md#git).

hhyyrylainen commented on src/microbe_stage/PatchMapGenerator.cs:
```diff
@@ -533,6 +533,7 @@ private static void BuildRegionSize(PatchRegion region)
             {
                 region.Width += offset;
                 region.Height += offset;
+                region.Columns = region.Rows = 1;
```
Does the columns and rows actually need to be stored like this? I get the impression that before the region patch positions was implicitly known based on the size, couldn't the same thing be done still (i.e. don't add the rows and columns variables but calculate them from the width and height of the region)?

hhyyrylainen commented on src/microbe_stage/editor/PatchMapDrawer.cs:
```diff
@@ -33,7 +36,7 @@ public partial class PatchMapDrawer : Control
     ///   The representation of connections between regions, so we won't draw the same connection multiple times
     /// </summary>
-    private readonly Dictionary<Vector2I, Vector2[]> connections = new();
+    private readonly Dictionary<Vector2I, (Vector2[], bool, bool)> connections = new();
```
I think this should use a named tuple here as otherwise it is very unclear what the two added booleans mean.

hhyyrylainen commented on src/microbe_stage/editor/PatchMapDrawer.cs:
```diff
@@ -384,13 +390,43 @@ private static Vector2 PatchCenter(Vector2 pos)
+            Width = width,
+            BeginCapMode = Line2D.LineCapMode.Round,
+            EndCapMode = Line2D.LineCapMode.Round,
+            ZIndex = zIndex,
```
ZIndex can mess with other layers than purely just this. For example even the pause menu might not work correctly. So I'd really prefer an approach where ZIndex is not used. Reordering the children in the line container can be used to set the draw order. If just a couple of ZIndex values are needed, they could be made as separate containers for the lines. Or what is it needed for? I can probably come up with some better suggestions if I know that.

hhyyrylainen commented on src/microbe_stage/editor/PatchMapDrawer.cs:
```diff
@@ -472,52 +503,166 @@ private void CreateRegionLinks()
+        switch (shape)
+        {
+            case "lower":
+                intermediate1 = new Vector2(startCenter.X, lower.End.Y + i * 50);
```
Case statements with multiple lines should be wrapped in braces according to our styleguide.

hhyyrylainen commented on src/microbe_stage/editor/PatchMapDrawer.cs:
```diff
@@ -892,22 +1041,47 @@ private void AddPatchNode(Patch patch, Vector2 position)
+            if (adjacent.Visibility == MapElementVisibility.Hidden ||
+                (adjacent.Visibility == MapElementVisibility.Unknown &&
+                    regionVisibility == MapElementVisibility.Unknown))
+                continue;
```
Multiline if statements should always use braces according to the styleguide.

hhyyrylainen commented on src/microbe_stage/editor/PatchMapDrawer.cs:
```diff
@@ -472,52 +503,166 @@ private void CreateRegionLinks()
         var startRect = new Rect2(start.ScreenCoordinates, start.Size);
         var endCenter = RegionCenter(end);
         var endRect = new Rect2(end.ScreenCoordinates, end.Size);
+        int priority;
```
Variables should be defined as close to where they are initialized as possible (so not all at the start of a method like ancient C styles mandated).

CountingBeeps commented:
This is a great start, but I don't always like how these look. I think the visuals could use a little work before this is merged. Specifically in this case:
![image](https://github.com/user-attachments/assets/6dd61160-7cda-40a9-9d03-041a538e7006)

hhyyrylainen commented:
I did a partial review of this PR. I left a big chunk of the code unreviewed but it was probably mostly just the path algorithm changes that caused the internal API to change, which looks long but isn't actually that complicated of a change...

This is already in pretty good shape I think overall design wise.

Anyway I'd like to first get the biggest issues out of the way before reviewing everything in here:
- The columns and rows variables which will break save compatibility
- Use of zindex in the line rendering on the map

Ones those are taken care of I think this will be close to done already. (though I just saw @CountingBeeps comment)

Patryk26g commented:
> This is a great start, but I don't always like how these look. I think the visuals could use a little work before this is merged. Specifically in this case: ![image](https://private-user-images.githubusercontent.com/60719255/401158876-6dd61160-7cda-40a9-9d03-041a538e7006.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzY2MTI0OTUsIm5iZiI6MTczNjYxMjE5NSwicGF0aCI6Ii82MDcxOTI1NS80MDExNTg4NzYtNmRkNjExNjAtN2NkYS00MGE5LTlkMDMtMDQxYTUzOGU3MDA2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTExVDE2MTYzNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQxNzRmN2Q2ODcwNGJiOGJhYjM0YjVhODYzMTI0ZmRhYjA5NDIxYzBjNjFhZWE0NmFmNTI2MDgxZDE3MjVkMDQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.rYrCmeRWAAWVyhSYjIEVJkvycJe2zRfYRIC_NuvMlQY)

It should look better now

Patryk26g commented:
> I did a partial review of this PR. I left a big chunk of the code unreviewed but it was probably mostly just the path algorithm changes that caused the internal API to change, which looks long but isn't actually that complicated of a change...
> 
> This is already in pretty good shape I think overall design wise.
> 
> Anyway I'd like to first get the biggest issues out of the way before reviewing everything in here:
> 
> * The columns and rows variables which will break save compatibility
> * Use of zindex in the line rendering on the map
> 
> Ones those are taken care of I think this will be close to done already. (though I just saw @CountingBeeps comment)

I have removed z-indeces and the column and row variables. I also adjusted the look of the patches

hhyyrylainen commented on src/microbe_stage/editor/PatchMapDrawer.cs:
```diff
@@ -472,52 +504,183 @@ private void CreateRegionLinks()
+        int priority = 0;
+        switch (shape)
+        {
+            case "lower":
```
I think this should be an enum as using strings with magic values like this isn't very good style.
Also the CI has found unused variables so, could you please fix those before I take a full new look at the code?

Patryk26g replied:
I updated the code

hhyyrylainen commented on src/microbe_stage/editor/PatchMapDrawer.cs:
```diff
@@ -384,15 +401,43 @@ private static Vector2 PatchCenter(Vector2 pos)
+        float halfLength = Constants.PATCH_NODE_RECT_LENGTH * 0.5f;
+        float posX = patchPos.X;
+        float posY = patchPos.Y;
+        List<Vector2> sideCenterPoints = new();
```
This method should be rewritten to not use this temporary memory. As we are working on a realtime game, memory allocations should be avoided whenever possible. Here it looks like it will be possible by writing more code to manually compare the values to get the actual single result value.

hhyyrylainen commented on src/microbe_stage/editor/PatchMapDrawer.cs:
```diff
@@ -472,52 +512,181 @@ private void CreateRegionLinks()
+        return priority;
+    }
+
+    private int GetIShapePathCharacteristics(int startRows, int startColumns, int endRows, int endColumns,
```
Shouldn't the name be `GetIShapePathPriority`? As this returns the priority and not some other unspecified characteristic.

Also the the other "characteristics" methods seem to have this same problem with them.

Patryk26g replied:
That is true but I wanted to be consistent in all these paths, but I will change it to priority

hhyyrylainen commented on src/microbe_stage/editor/PatchMapDrawer.cs:
```diff
@@ -472,52 +512,181 @@ private void CreateRegionLinks()
+        Vector2 intermediate1;
+        Vector2 intermediate2;
+        int priority = isTrueMiddle ? 2 : 1;
+        if (!flip)
```
I'd add a blank line here to give better readability for the if (as the variables above this aren't used in the if condition, I'd logically separate them with a blank line)

hhyyrylainen commented on src/microbe_stage/editor/PatchMapDrawer.cs:
```diff
@@ -472,52 +512,181 @@ private void CreateRegionLinks()
+        Vector2 intermediate1;
+        Vector2 intermediate2;
+        int priority = 0;
+        switch (side)
```
Same here with the empty line. I like to have room in the code for it to breathe rather than mushing all lines together like empty lines cost money.

hhyyrylainen commented on src/microbe_stage/editor/PatchMapDrawer.cs:
```diff
@@ -472,52 +512,181 @@ private void CreateRegionLinks()
         AdjustPathEndpoints();
     }
 
+    private int AdjustPriority(int size1, int priority)
```
For understandability I'd add a general comment before this group of priority methods for what the priority means, something like:

```
// The next few methods are for calculating path priorities. They are needed because ...
```

hhyyrylainen replied:
The change you did is slightly different in semantics, an XML comment *only* applies to the language element that it is before. So now this comment only applies to AdjustPriority. So I suggest it is reformatted to be a normal comment `//`  instead of `/// <summary>` and a blank line to be added after it and before AdjustPriority to make it clear the comment applies to multiple things and not just this one method.

hhyyrylainen commented on src/microbe_stage/editor/PatchMapDrawer.cs:
```diff
@@ -472,52 +512,181 @@ private void CreateRegionLinks()
+
+        const float offset = Constants.PATCH_NODE_RECT_LENGTH + Constants.PATCH_AND_REGION_MARGIN;
+        const float margin = Constants.PATCH_REGION_BORDER_WIDTH + Constants.PATCH_AND_REGION_MARGIN;
+        int startRows = (int)Math.Round((start.Height - margin) / offset);
+        var startColumns = (int)Math.Round((start.Width - margin) / offset);
+        int endRows = (int)Math.Round((end.Height - margin) / offset);
+        var endColumns = (int)Math.Round((end.Width - margin) / offset);
```
Why are int and var used inconsistently in this block of variables?

hhyyrylainen commented on src/microbe_stage/editor/PatchMapDrawer.cs:
```diff
@@ -526,45 +695,61 @@ private Vector2[] GetLeastIntersectingPath(PatchRegion start, PatchRegion end)
+        for (int i = -3; i <= 3; ++i)
+        {
+            var middlePoint =
+                new Vector2(left.End.X + right.Position.X + i * 25, upper.End.Y + lower.Position.Y + i * 25) / 2.0f;
```
The 25 here and the 3 few lines below seem like magic numbers I have no clue about, could you make those named const values in this class to clarify their reason why they are like that?

hhyyrylainen replied:
Also I know that the old code did this, but could you update any `/ 2.0f` pieces of code to be `* 0.5f` where they are safe to change? Floating point multiplies are faster to calculate than divides so it should be preferred whenever a divide is not absolutely required.

hhyyrylainen commented on src/microbe_stage/editor/PatchMapDrawer.cs:
```diff
@@ -899,15 +1086,48 @@ private void AddPatchNode(Patch patch, Vector2 position)
+            Vector2[] path = [PatchCenter(adjacent.ScreenCoordinates), closesPatchSideCenter, regionPoint];
+
+            // If the path is highlighted then create it later to be on top of other paths
+            if (!highlight)
```
Multiline ifs should always use braces according to our styleguide.

hhyyrylainen commented on src/microbe_stage/editor/PatchMapDrawer.cs:
```diff
@@ -917,15 +1137,18 @@ private void RebuildRegionConnections()
                 continue;
 
             // Check if connections should be highlighted
-            // (Unknown patches should not highlight connections)
```
What's the reason behind this change, making unknown patches now be able to highlight?

Patryk26g replied:
I think that it now makes more sense since it is connected anyway and user can click on it

Patryk26g replied:
![image](https://github.com/user-attachments/assets/32fbafd6-213b-4f42-9c30-8dbfaccda731)
Like, in my opinion if I click this patch then the connections should be highlighted even if they fade away

hhyyrylainen replied:
Yeah, that makes sense to me (the code was probably originally made for things to not look weird when something was partially visible). BTW that screenshot shows some aliasing artifacts, is the antialias option on for the lines already or not? If it isn't on already it might help a bit with the visuals.

hhyyrylainen reviewed (changes requested):
I think overall architecture-wise this is now in good shape. I did a full readthrough of the code and I commented on a few places that could do with some improved clarity or slight changes to match our styleguide. I didn't test in game yet but I liked how the visual screenshots looked. Even if the extra space looks unfamiliar, but to avoid like overlaps it looks like the empty space is required.

hhyyrylainen commented:
I've now tested this in-game and the results are really good.

I tried that antialias option and it doesn't really seem to do much, sadly. So I wouldn't turn it on.

As I tested this in-game I noticed there's some pretty unclear situations. In these kind of situations I think it would be more clear and hopefully easy to make the links between patches highlight whenever a patch is selected (so not just the between region links):

![kuva](https://github.com/user-attachments/assets/ab4a718d-358a-4b08-8798-aa9d2737be80)

Patryk26g commented:
Do not review it yet, I'm not finished

Patryk26g commented:
Ready for testing now

hhyyrylainen reviewed (approved):
I think the code is still probably fine since I last looked, and the visuals have been tweaked a lot already. There is one known problem but I'll open a separate issue for that after merging.

hhyyrylainen commented:
The follow up issue: https://github.com/Revolutionary-Games/Thrive/issues/5822
